Adding NIRISS SOSS notebook

rcooper295 opened the pull request:
This notebook checklist has been made available to us by the [the Notebooks For All team](https://github.com/Iota-School/notebooks-for-all/blob/main/resources/event-hackathon/notebook-authoring-checklist.md).
Its purpose is to serve as a guide for both the notebook author and the technical reviewer highlighting critical aspects to consider when striving to develop an accessible and effective notebook.

### The First Cell

- [x] The title of the notebook in a first-level heading (eg. `<h1>` or `# in markdown`).
- [x] A brief description of the notebook.
- [x] A table of contents in an [ordered list](https://www.markdownguide.org/basic-syntax/#ordered-lists) (`1., 2.,` etc. in Markdown).
- [x] The author(s) and affiliation(s) (if relevant).
- [ ] The date first published.
- [x] The date last edited (if relevant).
- [ ] A link to the notebook's source(s) (if relevant).

### The Rest of the Cells

- [x] There is only one H1 (`#` in Markdown) used in the notebook.
- [x] The notebook uses other [heading tags](https://www.markdownguide.org/basic-syntax/#headings) in order (meaning it does not skip numbers). 

## Text

- [x] All link text is descriptive. It tells users where they will be taken if they open the link.
- [ ] All acronyms are defined at least the first time they are used. 
- [x] Field-specific/specialized terms are used when needed, but not excessively.

## Code

- [x] Code sections are introduced and explained before they appear in the notebook. This can be fulfilled with a heading in a prior Markdown cell, a sentence preceding it, or a code comment in the code section.
- [x] Code has explanatory comments (if relevant). This is most important for long sections of code.
- [ ] If the author has control over the syntax highlighting theme in the notebook, that theme has enough color contrast to be legible.
- [ ] Code and code explanations focus on one task at a time. Unless comparison is the point of the notebook, only one method for completing the task is described at a time.

### Images

- [ ] All images (jpg, png, svgs) have an [image description](https://www.w3.org/WAI/tutorials/images/decision-tree/). This could be
    - [ ] [Alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt) (an `alt` property) 
    - [ ] [Empty alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt#decorative_images) for decorative images/images meant to be skipped (an `alt` attribute with no value)
    - [ ] Captions
    - [ ] If no other options will work, the image is decribed in surrounding paragraphs.

- [ ] Any [text present in images](https://www.w3.org/WAI/WCAG21/Understanding/images-of-text.html) exists in a text form outside of the image (this can be alt text, captions, or surrounding text.)

### Visualizations

- [x] All visualizations have an image description. Review the previous section, Images, for more information on how to add it.
- [x] [Visualization descriptions](http://diagramcenter.org/specific-guidelines-e.html) include
    - [ ] The type of visualization (like bar chart, scatter plot, etc.)
    - [ ] Title
    - [ ] Axis labels and range
    - [ ] Key or legend
    - [ ] An explanation of the visualization's significance to the notebook (like the trend, an outlier in the data, what the author learned from it, etc.)

- [ ] All visualizations and their parts have [enough color contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) ([color contrast checker](https://webaim.org/resources/contrastchecker/)) to be legible. Remember that transparent colors have lower contrast than their opaque versions.
- [ ] All visualizations [convey information with more visual cues than color coding](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html). Use text labels, patterns, or icons alongside color to achieve this.
- [ ] All visualizations have an additional way for notebook readers to access the information. Linking to the original data, including a table of the data in the same notebook, or sonifying the plot are all options.

rcooper295 committed:
first commit of SOSS notebook

rcooper295 committed:
pep8 fixes

rcooper295 commented:
Notebook has been reviewed internally by NIRISS SOSS team and is now ready for review by repo maintainers.

drlaw1558 committed:
Merge branch 'main' into sossnb

drlaw1558 commented:
Prior to digging into details of this notebook, two quick questions for @rcooper295 :

1) Presumably the package dependency that this needs is 'batman-package', not 'batman'?  Ordinarily we'd try not to include unusual dependencies like this, although in this case it appears to be a major package that's key to working with SOSS data.  'batman' is something completely different though on Pypi, and last updated in 2013.

2) This notebook really drives home the importance of the issue being worked at the moment with SOSS runtime.  I've had the notebook running for a few hours, and it's still less than half-way through spec2 spectral extraction.  Can we replace the example data set from PID 1366 with one that's less time-consuming?  At the moment I don't think CI on this notebook will be plausible given the wall-clock limitations of the github runners (even the larger paid runners).  Similarly, for folks new to the mode it would be be much easier if the demo case didn't take a day to run.  Is there a different data set with significantly fewer integrations that we could use instead?  At least until such time as the pipeline is made faster?

One final bonus comment: In the TSO3 setup right now the association generation hard-codes PID=1366 and name='WASP 39b'.  That limits the ability of this notebook to be easily run on other data sets; can we please populate this information instead from the headers of the files being processed?

bhilbert4 commented:
Just a quick comment from someone who also has a package that depends on batman-package: I was having trouble getting batman to work in python 3.12 (perhaps related to [this](https://github.com/lkreidberg/batman/issues/80) Issue).  Also, I know it's an edge case at this point, but the latest release of batman [does not work](https://github.com/lkreidberg/batman/issues/79) for those running on Macs with Intel chips. In that case, users are forced to use the previous release, but that release requires numpy<2, which is incompatible with a lot of other dependencies, including the jwst package.

rcooper295 commented:
@drlaw1558 Thanks for your questions. 
1. Yes, that should have been `batman-package` instead. It's pretty widely used, and while the model lightcurve could theoretically be omitted from this notebook, I do think it is useful to show. However if it poses a major headache because of the dependency issues Bryan mentioned I can remove it.
2. Runtime is definitely an issue; I had to run the spec2 and tso3 portions of this notebook in a tmux session on a witserv machine to keep it from halting when my computer went to sleep, and of course we can't count on all potential users having access to a similar setup. I'll have to talk with the SOSS leads to see if there's another suitable dataset with fewer integrations that we could use for this notebook, though this ERS program produced one of the first major science results of JWST and is probably therefore already familiar to a lot of users, which I think is helpful for lowering the barriers to entry. 

I can make the change to populate the name and PID for the association from the headers rather than hardcoding them.

rcooper295 commented:
Hi @drlaw1558, I spoke with the SOSS team and they came up with a possible alternative dataset we could use: PID 2734 (WASP-96), which has 280 integrations, so it should cut the runtime down by as much as half. I wanted to check with you that that is a sufficient improvement before I rewrite the program-specific parts of the notebook. Otherwise, we could release the notebook as-is with a caveat about the long runtime, and then just update it if/when the spectral extraction steps are eventually sped up.

drlaw1558 commented:
Hi @rcooper295 ; using PID 2734 should help, let's do that.  We may have to just live with failing CI on this notebook for the next few months until we get some speed improvements into the pipeline.

Another thing to consider would be using only a single segment, but I'd imagine that would be very undesirable as you wouldn't have a normal-looking light curve, and may miss the transit signal entirely.

rcooper295 committed:
Updated notebook to use PID 2734, WASP-96b transit since it can be processed more quickly

drlaw1558 committed:
Generalization, requirements update

drlaw1558 committed:
PEP8, batman note

drlaw1558 commented:
@rcooper295 I've made a couple of additional modifications that should be uncontroversial.  I've updated it to give the latest pipeline version (so we don't need to update it again next week), fix batman to batman-package, and added a new sort_files function to determine which files are TA, dispersed, and F277W files.  This way the notebook won't break if someone tries running it on data that doesn't have TA files in the directory, if there are different numbers of exposures, etc.

I've also added a brief note about the architecture dependence of the batman package to the intro section in case anyone runs into any trouble with it.

Let me know what you think of these changes, otherwise I think we're about set to merge it so that I can try out the CI.

rcooper295 committed:
small speedup of sort_files, text updates

rcooper295 commented:
@drlaw1558 Those changes look good, thanks! I did make a few small changes: opening the datamodels an extra time in the `sort_files()` function slowed things down considerably when I ran it, so while I generally tried to use the datamodels throughout the notebook for demo purposes, I did resort to using `astropy.io.fits` functions when I only needed to access header keywords. I modified your function to use `fits.get_val()` instead, but let me know if you'd prefer to use the datamodels anyways. 
There were a few other very minor changes (a debugging cell I deleted, a slight modification to the batman disclaimer) that I made but hopefully it's nearly there!

rcooper295 committed:
pep8

drlaw1558 committed:
Infrastructure

drlaw1558 committed:
Update citation list

drlaw1558 commented:
Fixed one last typo, added new authors to the citation file, added the infrastructure to include this notebook in our rendered TOC and README, and added to the larger runners list given the size of the products.  I think this is ready to merge, and we can address CI issues with the long runtime in 12.1 as they arise.  Thanks for putting this together @rcooper295 !
